### ISSUE TITLE
Processing fails when a search returns 25 text values with standoff

### ISSUE THREAD

**Issue by benjamingeer** [dhlab-basel/Knora#913]
Try this on the `develop` branch with the BEOL data loaded:

http://localhost:3333/v2/search/Hochwohlgebohrner?limitToStandoffClass=http%3A%2F%2Fapi.knora.org%2Fontology%2Fstandoff%2Fsimple%2Fv2%23StandoffParagraphTag

The response:

```
{
"error": "java.util.NoSuchElementException: key not found: http://www.knora.org/ontology/knora-base#standoffTagHasEnd"
}
```

The stack trace on the console:

```
java.util.NoSuchElementException: key not found: http://www.knora.org/ontology/knora-base#standoffTagHasEnd
	at scala.collection.immutable.Map$Map2.apply(Map.scala:129)
	at org.knora.webapi.util.standoff.StandoffTagUtilV2$.$anonfun$createStandoffTagsV2FromSparqlResults$1(StandoffTagUtilV2.scala:769)
	at org.knora.webapi.util.standoff.StandoffTagUtilV2$$$Lambda$1584/1506996791.apply(Unknown Source)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$Lambda$12/1676605578.apply(Unknown Source)
	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:231)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
	at scala.collection.TraversableLike.map(TraversableLike.scala:234)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.knora.webapi.util.standoff.StandoffTagUtilV2$.createStandoffTagsV2FromSparqlResults(StandoffTagUtilV2.scala:706)
	at org.knora.webapi.util.ConstructResponseUtilV2$.createValueContentV2FromValueRdfData(ConstructResponseUtilV2.scala:463)
	at org.knora.webapi.util.ConstructResponseUtilV2$.$anonfun$constructReadResourceV2$3(ConstructResponseUtilV2.scala:686)
	at org.knora.webapi.util.ConstructResponseUtilV2$$$Lambda$1560/390556254.apply(Unknown Source)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$Lambda$12/1676605578.apply(Unknown Source)
	at scala.collection.Iterator.foreach(Iterator.scala:929)
	at scala.collection.Iterator.foreach$(Iterator.scala:929)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
	at scala.collection.IterableLike.foreach(IterableLike.scala:71)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:70)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableLike.map(TraversableLike.scala:234)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:227)
	at scala.collection.AbstractTraversable.map(Traversable.scala:104)
	at org.knora.webapi.util.ConstructResponseUtilV2$.$anonfun$constructReadResourceV2$1(ConstructResponseUtilV2.scala:684)
```

This is because this simple query, for just 25 resources and 25 values, returns more than 50000 statements. The response from the triplestore is truncated, so it ends with an incomplete standoff class, which is why `ConstructResponseUtilV2` throws an exception.

(Moved from #912.)

**Comment by benjamingeer** [comment]
Or just this:

http://localhost:3333/v2/search/Hochwohlgebohrner

```
{
  "error": "java.util.NoSuchElementException: key not found: http://www.knora.org/ontology/knora-base#standoffTagHasUUID"
}
```

**Comment by tobiasschweizer** [comment]
@benjamingeer So this means we have to find a configuration setting for `owlim:query-limit-results` that is never exceeded, but still protects us from attacks?

**Comment by tobiasschweizer** [comment]
I could also try to optimize the `CONSTRUCT` queries so we do not get information that we do not need.

**Comment by benjamingeer** [comment]
> I could also try to optimize the `CONSTRUCT` queries so we do not get information that we do not need.

That would be great. We could also:

* Set `results-per-page` to something smaller, like 5 instead of 25.
* More gracefully handle the situation where we get an incomplete result because it was truncated by the triplestore.
* Consider changing the way we store standoff, so it requires fewer triples.

**Comment by benjamingeer** [comment]
```sparql
CONSTRUCT {
?resource <http://www.knora.org/ontology/knora-base#isMainResource> "true"^^<http://www.w3.org/2001/XMLSchema#boolean> .
?resource <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
?resource ?resourceProp ?resourceObj .
?resource <http://www.knora.org/ontology/knora-base#hasValue> ?resourceValueObject .
?resource ?resourceValueProp ?resourceValueObject .
?resourceValueObject ?resourceValueObjectProp ?resourceValueObjectObj .
?resourceValueObject <http://www.knora.org/ontology/knora-base#valueHasStandoff> ?standoffNode .
?standoffNode ?standoffProp ?standoffValue .
} WHERE {
{
VALUES ?resource { <http://rdfh.ch/0801/beol/Ou0NrqS8T-WY7lHTbkUKww> <http://rdfh.ch/0801/beol/4Wfnmh7uQ522AH9OWEbygA> <http://rdfh.ch/0801/beol/2PeTcAP2R0uyhxKMR163pA> <http://rdfh.ch/0801/beol/6vuGLYDSQfqUBW8LOmMoaw> <http://rdfh.ch/0801/beol/CyrFhLONSVqyn0hMe1PyAA> <http://rdfh.ch/0801/beol/M-ZlPGzRTV2Qjye8ynw1QQ> <http://rdfh.ch/0801/beol/M7TWHhotSTiXGD9YSPZkNA> <http://rdfh.ch/0801/beol/M8UMKPYOSqiLG4oqunxr-Q> <http://rdfh.ch/0801/beol/N80rnvmsTeuTF1QnLfdC2w> <http://rdfh.ch/0801/beol/OIUG-7dnReCskHcKecGU6w> <http://rdfh.ch/0801/beol/45jLFi29Rn-xl2wO8lCbkg> <http://rdfh.ch/0801/beol/0VE-1W0XRB-vzmZV2aemnA> <http://rdfh.ch/0801/beol/8b6a0T01T4iOODgOGDDPcQ> <http://rdfh.ch/0801/beol/4auvM0v_TzKS4j2OVL8RiA> <http://rdfh.ch/0801/beol/F_aBo9KCSJyRIGXXTLta5w> <http://rdfh.ch/0801/beol/QewXbcXNTZOzPuKyEirsiw> <http://rdfh.ch/0801/beol/QaDxYckoSY-Mu5tU8Gbz4w> <http://rdfh.ch/0801/beol/FiFZoNxOTvumQJSLonYp0A> <http://rdfh.ch/0801/beol/IBu4vDCGSTm4G3TNsOOtyA> <http://rdfh.ch/0801/beol/PJDKzIXlRomYrcsJ3bvDOg> <http://rdfh.ch/0801/beol/LPclPw5MTm2jS_W7OE2ktw> <http://rdfh.ch/0801/beol/3tct2_3BRLmnSoA45OHBAA> <http://rdfh.ch/0801/beol/-0tI3HXgSSOeDtkf-SA00w> <http://rdfh.ch/0801/beol/5V0UQ8nyQquHXk5rdiWJ-Q> <http://rdfh.ch/0801/beol/F6gjeBItSeS2bDGzmBe41w> }
?resource <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
GRAPH <http://www.ontotext.com/explicit> {
    ?resource <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
}
GRAPH <http://www.ontotext.com/explicit> {
    ?resource ?resourceProp ?resourceObj .
}
} UNION {
VALUES ?resourceValueObject { <http://rdfh.ch/0801/beol/4auvM0v_TzKS4j2OVL8RiA/values/0W3GdDbtTgKfXR_jI62fHQ> <http://rdfh.ch/0801/beol/45jLFi29Rn-xl2wO8lCbkg/values/RM8btSjaQWepuwFnQ-s0Yw> <http://rdfh.ch/0801/beol/M7TWHhotSTiXGD9YSPZkNA/values/q8sJbBrnTSmHyswzK7vhLw> <http://rdfh.ch/0801/beol/-0tI3HXgSSOeDtkf-SA00w/values/GJQB9IHYTl2RhG4g4ru0YA> <http://rdfh.ch/0801/beol/LPclPw5MTm2jS_W7OE2ktw/values/gylS9v1hTfWH6_u6HfgU0Q> <http://rdfh.ch/0801/beol/Ou0NrqS8T-WY7lHTbkUKww/values/7iAa9ZXLTAqLCBp5yNKh3Q> <http://rdfh.ch/0801/beol/N80rnvmsTeuTF1QnLfdC2w/values/xbee-ZCRTE2XN6JURHcUqg> <http://rdfh.ch/0801/beol/6vuGLYDSQfqUBW8LOmMoaw/values/omx8e3xMSMKfUl3SAKJVow> <http://rdfh.ch/0801/beol/F6gjeBItSeS2bDGzmBe41w/values/lgKU4Hu9QyisG6TTwnr62g> <http://rdfh.ch/0801/beol/2PeTcAP2R0uyhxKMR163pA/values/NPWvNLAKTz-OOPbrXaH6Sg> <http://rdfh.ch/0801/beol/M8UMKPYOSqiLG4oqunxr-Q/values/mgC-KayaSYGHp6oYWqu2mQ> <http://rdfh.ch/0801/beol/QaDxYckoSY-Mu5tU8Gbz4w/values/QnP-RUaKTRWGzEayii8BfA> <http://rdfh.ch/0801/beol/5V0UQ8nyQquHXk5rdiWJ-Q/values/dBMQsL-IQNeIhuqke2Vw4Q> <http://rdfh.ch/0801/beol/0VE-1W0XRB-vzmZV2aemnA/values/SSMaDrOoRXSUaNMXQFEfZQ> <http://rdfh.ch/0801/beol/FiFZoNxOTvumQJSLonYp0A/values/q7NPps2UTIunu5j1WTLb7w> <http://rdfh.ch/0801/beol/CyrFhLONSVqyn0hMe1PyAA/values/zwa__RbFTUG9k3Pxg4R4gQ> <http://rdfh.ch/0801/beol/QewXbcXNTZOzPuKyEirsiw/values/aZWbXo1VSxOcByMm7bL79Q> <http://rdfh.ch/0801/beol/OIUG-7dnReCskHcKecGU6w/values/IINpkmISQr2ecIwRs9IRaA> <http://rdfh.ch/0801/beol/PJDKzIXlRomYrcsJ3bvDOg/values/fdblBD43SYelpS_Xr1mUxQ> <http://rdfh.ch/0801/beol/IBu4vDCGSTm4G3TNsOOtyA/values/epaW8kuXT4qsc4ZuFE9ZGw> <http://rdfh.ch/0801/beol/M-ZlPGzRTV2Qjye8ynw1QQ/values/E3yHIBWtQ8SGjfuqQx5HDw> <http://rdfh.ch/0801/beol/4Wfnmh7uQ522AH9OWEbygA/values/Sss4_gFqTvefnkmI027TnQ> <http://rdfh.ch/0801/beol/F_aBo9KCSJyRIGXXTLta5w/values/LsHcTGTVTIevf9K6NHOT5Q> <http://rdfh.ch/0801/beol/8b6a0T01T4iOODgOGDDPcQ/values/xYV-JzpdSU-JDNrdGXl_9w> <http://rdfh.ch/0801/beol/3tct2_3BRLmnSoA45OHBAA/values/VcA57tXxSbKn8muHHbyYCg> }
?resource <http://www.knora.org/ontology/knora-base#hasValue> ?resourceValueObject .
GRAPH <http://www.ontotext.com/explicit> {
    ?resource ?resourceValueProp ?resourceValueObject .
}
GRAPH <http://www.ontotext.com/explicit> {
    ?resourceValueObject <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
}
GRAPH <http://www.ontotext.com/explicit> {
    ?resourceValueObject ?resourceValueObjectProp ?resourceValueObjectObj .
}
} UNION {
VALUES ?resourceValueObject { <http://rdfh.ch/0801/beol/4auvM0v_TzKS4j2OVL8RiA/values/0W3GdDbtTgKfXR_jI62fHQ> <http://rdfh.ch/0801/beol/45jLFi29Rn-xl2wO8lCbkg/values/RM8btSjaQWepuwFnQ-s0Yw> <http://rdfh.ch/0801/beol/M7TWHhotSTiXGD9YSPZkNA/values/q8sJbBrnTSmHyswzK7vhLw> <http://rdfh.ch/0801/beol/-0tI3HXgSSOeDtkf-SA00w/values/GJQB9IHYTl2RhG4g4ru0YA> <http://rdfh.ch/0801/beol/LPclPw5MTm2jS_W7OE2ktw/values/gylS9v1hTfWH6_u6HfgU0Q> <http://rdfh.ch/0801/beol/Ou0NrqS8T-WY7lHTbkUKww/values/7iAa9ZXLTAqLCBp5yNKh3Q> <http://rdfh.ch/0801/beol/N80rnvmsTeuTF1QnLfdC2w/values/xbee-ZCRTE2XN6JURHcUqg> <http://rdfh.ch/0801/beol/6vuGLYDSQfqUBW8LOmMoaw/values/omx8e3xMSMKfUl3SAKJVow> <http://rdfh.ch/0801/beol/F6gjeBItSeS2bDGzmBe41w/values/lgKU4Hu9QyisG6TTwnr62g> <http://rdfh.ch/0801/beol/2PeTcAP2R0uyhxKMR163pA/values/NPWvNLAKTz-OOPbrXaH6Sg> <http://rdfh.ch/0801/beol/M8UMKPYOSqiLG4oqunxr-Q/values/mgC-KayaSYGHp6oYWqu2mQ> <http://rdfh.ch/0801/beol/QaDxYckoSY-Mu5tU8Gbz4w/values/QnP-RUaKTRWGzEayii8BfA> <http://rdfh.ch/0801/beol/5V0UQ8nyQquHXk5rdiWJ-Q/values/dBMQsL-IQNeIhuqke2Vw4Q> <http://rdfh.ch/0801/beol/0VE-1W0XRB-vzmZV2aemnA/values/SSMaDrOoRXSUaNMXQFEfZQ> <http://rdfh.ch/0801/beol/FiFZoNxOTvumQJSLonYp0A/values/q7NPps2UTIunu5j1WTLb7w> <http://rdfh.ch/0801/beol/CyrFhLONSVqyn0hMe1PyAA/values/zwa__RbFTUG9k3Pxg4R4gQ> <http://rdfh.ch/0801/beol/QewXbcXNTZOzPuKyEirsiw/values/aZWbXo1VSxOcByMm7bL79Q> <http://rdfh.ch/0801/beol/OIUG-7dnReCskHcKecGU6w/values/IINpkmISQr2ecIwRs9IRaA> <http://rdfh.ch/0801/beol/PJDKzIXlRomYrcsJ3bvDOg/values/fdblBD43SYelpS_Xr1mUxQ> <http://rdfh.ch/0801/beol/IBu4vDCGSTm4G3TNsOOtyA/values/epaW8kuXT4qsc4ZuFE9ZGw> <http://rdfh.ch/0801/beol/M-ZlPGzRTV2Qjye8ynw1QQ/values/E3yHIBWtQ8SGjfuqQx5HDw> <http://rdfh.ch/0801/beol/4Wfnmh7uQ522AH9OWEbygA/values/Sss4_gFqTvefnkmI027TnQ> <http://rdfh.ch/0801/beol/F_aBo9KCSJyRIGXXTLta5w/values/LsHcTGTVTIevf9K6NHOT5Q> <http://rdfh.ch/0801/beol/8b6a0T01T4iOODgOGDDPcQ/values/xYV-JzpdSU-JDNrdGXl_9w> <http://rdfh.ch/0801/beol/3tct2_3BRLmnSoA45OHBAA/values/VcA57tXxSbKn8muHHbyYCg> }
GRAPH <http://www.ontotext.com/explicit> {
    ?resourceValueObject <http://www.knora.org/ontology/knora-base#valueHasStandoff> ?standoffNode .
}
GRAPH <http://www.ontotext.com/explicit> {
    ?standoffNode ?standoffProp ?standoffValue .
}
}
}
```

**Comment by benjamingeer** [comment]
After discussion, we'd like to try returning less standoff in search results. We could just return the following:

* If a standoff tag was searched for, return that tag and its parent tag (but not its siblings or any other tags). Perhaps even not return all of the text.
* If no standoff tag was searched for, don't return any markup, just return plain text.

**Comment by tobiasschweizer** [comment]
The method that creates the main query could accept a set of standoff node Iris as a parameter and would only return those, if any are specified (set could be emtpy too)

**Comment by benjamingeer** [comment]
I've just tried modifying the example `CONSTRUCT` query above to return the basic standoff data in a more compact form, and it seems to work well:

``` sparql
CONSTRUCT {
  ?resource <http://www.knora.org/ontology/knora-base#isMainResource> "true"^^<http://www.w3.org/2001/XMLSchema#boolean> .
  ?resource <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
  ?resource ?resourceProp ?resourceObj .
  ?resource <http://www.knora.org/ontology/knora-base#hasValue> ?resourceValueObject .
  ?resource ?resourceValueProp ?resourceValueObject .
  ?resourceValueObject ?resourceValueObjectProp ?resourceValueObjectObj .
  ?resourceValueObject <http://www.knora.org/ontology/knora-base#valueHasStandoff> ?standoffNode .
  ?standoffNode <http://www.knora.org/ontology/knora-base#standoffNodeHasData> ?standoffNodeData .
  ?standoffNode ?standoffProp ?standoffValue .
} WHERE {
  {
    VALUES ?resource { <http://rdfh.ch/0801/beol/Ou0NrqS8T-WY7lHTbkUKww> <http://rdfh.ch/0801/beol/4Wfnmh7uQ522AH9OWEbygA> <http://rdfh.ch/0801/beol/2PeTcAP2R0uyhxKMR163pA> <http://rdfh.ch/0801/beol/6vuGLYDSQfqUBW8LOmMoaw> <http://rdfh.ch/0801/beol/CyrFhLONSVqyn0hMe1PyAA> <http://rdfh.ch/0801/beol/M-ZlPGzRTV2Qjye8ynw1QQ> <http://rdfh.ch/0801/beol/M7TWHhotSTiXGD9YSPZkNA> <http://rdfh.ch/0801/beol/M8UMKPYOSqiLG4oqunxr-Q> <http://rdfh.ch/0801/beol/N80rnvmsTeuTF1QnLfdC2w> <http://rdfh.ch/0801/beol/OIUG-7dnReCskHcKecGU6w> <http://rdfh.ch/0801/beol/45jLFi29Rn-xl2wO8lCbkg> <http://rdfh.ch/0801/beol/0VE-1W0XRB-vzmZV2aemnA> <http://rdfh.ch/0801/beol/8b6a0T01T4iOODgOGDDPcQ> <http://rdfh.ch/0801/beol/4auvM0v_TzKS4j2OVL8RiA> <http://rdfh.ch/0801/beol/F_aBo9KCSJyRIGXXTLta5w> <http://rdfh.ch/0801/beol/QewXbcXNTZOzPuKyEirsiw> <http://rdfh.ch/0801/beol/QaDxYckoSY-Mu5tU8Gbz4w> <http://rdfh.ch/0801/beol/FiFZoNxOTvumQJSLonYp0A> <http://rdfh.ch/0801/beol/IBu4vDCGSTm4G3TNsOOtyA> <http://rdfh.ch/0801/beol/PJDKzIXlRomYrcsJ3bvDOg> <http://rdfh.ch/0801/beol/LPclPw5MTm2jS_W7OE2ktw> <http://rdfh.ch/0801/beol/3tct2_3BRLmnSoA45OHBAA> <http://rdfh.ch/0801/beol/-0tI3HXgSSOeDtkf-SA00w> <http://rdfh.ch/0801/beol/5V0UQ8nyQquHXk5rdiWJ-Q> <http://rdfh.ch/0801/beol/F6gjeBItSeS2bDGzmBe41w> }
    ?resource <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.knora.org/ontology/knora-base#Resource> .
    GRAPH <http://www.ontotext.com/explicit> {
      ?resource <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
    }
    GRAPH <http://www.ontotext.com/explicit> {
      ?resource ?resourceProp ?resourceObj .
    }
  } UNION {
    VALUES ?resourceValueObject { <http://rdfh.ch/0801/beol/4auvM0v_TzKS4j2OVL8RiA/values/0W3GdDbtTgKfXR_jI62fHQ> <http://rdfh.ch/0801/beol/45jLFi29Rn-xl2wO8lCbkg/values/RM8btSjaQWepuwFnQ-s0Yw> <http://rdfh.ch/0801/beol/M7TWHhotSTiXGD9YSPZkNA/values/q8sJbBrnTSmHyswzK7vhLw> <http://rdfh.ch/0801/beol/-0tI3HXgSSOeDtkf-SA00w/values/GJQB9IHYTl2RhG4g4ru0YA> <http://rdfh.ch/0801/beol/LPclPw5MTm2jS_W7OE2ktw/values/gylS9v1hTfWH6_u6HfgU0Q> <http://rdfh.ch/0801/beol/Ou0NrqS8T-WY7lHTbkUKww/values/7iAa9ZXLTAqLCBp5yNKh3Q> <http://rdfh.ch/0801/beol/N80rnvmsTeuTF1QnLfdC2w/values/xbee-ZCRTE2XN6JURHcUqg> <http://rdfh.ch/0801/beol/6vuGLYDSQfqUBW8LOmMoaw/values/omx8e3xMSMKfUl3SAKJVow> <http://rdfh.ch/0801/beol/F6gjeBItSeS2bDGzmBe41w/values/lgKU4Hu9QyisG6TTwnr62g> <http://rdfh.ch/0801/beol/2PeTcAP2R0uyhxKMR163pA/values/NPWvNLAKTz-OOPbrXaH6Sg> <http://rdfh.ch/0801/beol/M8UMKPYOSqiLG4oqunxr-Q/values/mgC-KayaSYGHp6oYWqu2mQ> <http://rdfh.ch/0801/beol/QaDxYckoSY-Mu5tU8Gbz4w/values/QnP-RUaKTRWGzEayii8BfA> <http://rdfh.ch/0801/beol/5V0UQ8nyQquHXk5rdiWJ-Q/values/dBMQsL-IQNeIhuqke2Vw4Q> <http://rdfh.ch/0801/beol/0VE-1W0XRB-vzmZV2aemnA/values/SSMaDrOoRXSUaNMXQFEfZQ> <http://rdfh.ch/0801/beol/FiFZoNxOTvumQJSLonYp0A/values/q7NPps2UTIunu5j1WTLb7w> <http://rdfh.ch/0801/beol/CyrFhLONSVqyn0hMe1PyAA/values/zwa__RbFTUG9k3Pxg4R4gQ> <http://rdfh.ch/0801/beol/QewXbcXNTZOzPuKyEirsiw/values/aZWbXo1VSxOcByMm7bL79Q> <http://rdfh.ch/0801/beol/OIUG-7dnReCskHcKecGU6w/values/IINpkmISQr2ecIwRs9IRaA> <http://rdfh.ch/0801/beol/PJDKzIXlRomYrcsJ3bvDOg/values/fdblBD43SYelpS_Xr1mUxQ> <http://rdfh.ch/0801/beol/IBu4vDCGSTm4G3TNsOOtyA/values/epaW8kuXT4qsc4ZuFE9ZGw> <http://rdfh.ch/0801/beol/M-ZlPGzRTV2Qjye8ynw1QQ/values/E3yHIBWtQ8SGjfuqQx5HDw> <http://rdfh.ch/0801/beol/4Wfnmh7uQ522AH9OWEbygA/values/Sss4_gFqTvefnkmI027TnQ> <http://rdfh.ch/0801/beol/F_aBo9KCSJyRIGXXTLta5w/values/LsHcTGTVTIevf9K6NHOT5Q> <http://rdfh.ch/0801/beol/8b6a0T01T4iOODgOGDDPcQ/values/xYV-JzpdSU-JDNrdGXl_9w> <http://rdfh.ch/0801/beol/3tct2_3BRLmnSoA45OHBAA/values/VcA57tXxSbKn8muHHbyYCg> }
    ?resource <http://www.knora.org/ontology/knora-base#hasValue> ?resourceValueObject .
    GRAPH <http://www.ontotext.com/explicit> {
      ?resource ?resourceValueProp ?resourceValueObject .
    }
    GRAPH <http://www.ontotext.com/explicit> {
      ?resourceValueObject <http://www.knora.org/ontology/knora-base#isDeleted> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .
    }
    GRAPH <http://www.ontotext.com/explicit> {
      ?resourceValueObject ?resourceValueObjectProp ?resourceValueObjectObj .
    }
  } UNION {
    VALUES ?resourceValueObject { <http://rdfh.ch/0801/beol/4auvM0v_TzKS4j2OVL8RiA/values/0W3GdDbtTgKfXR_jI62fHQ> <http://rdfh.ch/0801/beol/45jLFi29Rn-xl2wO8lCbkg/values/RM8btSjaQWepuwFnQ-s0Yw> <http://rdfh.ch/0801/beol/M7TWHhotSTiXGD9YSPZkNA/values/q8sJbBrnTSmHyswzK7vhLw> <http://rdfh.ch/0801/beol/-0tI3HXgSSOeDtkf-SA00w/values/GJQB9IHYTl2RhG4g4ru0YA> <http://rdfh.ch/0801/beol/LPclPw5MTm2jS_W7OE2ktw/values/gylS9v1hTfWH6_u6HfgU0Q> <http://rdfh.ch/0801/beol/Ou0NrqS8T-WY7lHTbkUKww/values/7iAa9ZXLTAqLCBp5yNKh3Q> <http://rdfh.ch/0801/beol/N80rnvmsTeuTF1QnLfdC2w/values/xbee-ZCRTE2XN6JURHcUqg> <http://rdfh.ch/0801/beol/6vuGLYDSQfqUBW8LOmMoaw/values/omx8e3xMSMKfUl3SAKJVow> <http://rdfh.ch/0801/beol/F6gjeBItSeS2bDGzmBe41w/values/lgKU4Hu9QyisG6TTwnr62g> <http://rdfh.ch/0801/beol/2PeTcAP2R0uyhxKMR163pA/values/NPWvNLAKTz-OOPbrXaH6Sg> <http://rdfh.ch/0801/beol/M8UMKPYOSqiLG4oqunxr-Q/values/mgC-KayaSYGHp6oYWqu2mQ> <http://rdfh.ch/0801/beol/QaDxYckoSY-Mu5tU8Gbz4w/values/QnP-RUaKTRWGzEayii8BfA> <http://rdfh.ch/0801/beol/5V0UQ8nyQquHXk5rdiWJ-Q/values/dBMQsL-IQNeIhuqke2Vw4Q> <http://rdfh.ch/0801/beol/0VE-1W0XRB-vzmZV2aemnA/values/SSMaDrOoRXSUaNMXQFEfZQ> <http://rdfh.ch/0801/beol/FiFZoNxOTvumQJSLonYp0A/values/q7NPps2UTIunu5j1WTLb7w> <http://rdfh.ch/0801/beol/CyrFhLONSVqyn0hMe1PyAA/values/zwa__RbFTUG9k3Pxg4R4gQ> <http://rdfh.ch/0801/beol/QewXbcXNTZOzPuKyEirsiw/values/aZWbXo1VSxOcByMm7bL79Q> <http://rdfh.ch/0801/beol/OIUG-7dnReCskHcKecGU6w/values/IINpkmISQr2ecIwRs9IRaA> <http://rdfh.ch/0801/beol/PJDKzIXlRomYrcsJ3bvDOg/values/fdblBD43SYelpS_Xr1mUxQ> <http://rdfh.ch/0801/beol/IBu4vDCGSTm4G3TNsOOtyA/values/epaW8kuXT4qsc4ZuFE9ZGw> <http://rdfh.ch/0801/beol/M-ZlPGzRTV2Qjye8ynw1QQ/values/E3yHIBWtQ8SGjfuqQx5HDw> <http://rdfh.ch/0801/beol/4Wfnmh7uQ522AH9OWEbygA/values/Sss4_gFqTvefnkmI027TnQ> <http://rdfh.ch/0801/beol/F_aBo9KCSJyRIGXXTLta5w/values/LsHcTGTVTIevf9K6NHOT5Q> <http://rdfh.ch/0801/beol/8b6a0T01T4iOODgOGDDPcQ/values/xYV-JzpdSU-JDNrdGXl_9w> <http://rdfh.ch/0801/beol/3tct2_3BRLmnSoA45OHBAA/values/VcA57tXxSbKn8muHHbyYCg> }
    GRAPH <http://www.ontotext.com/explicit> {
      ?resourceValueObject <http://www.knora.org/ontology/knora-base#valueHasStandoff> ?standoffNode .
    }
    GRAPH <http://www.ontotext.com/explicit> {
      ?standoffNode <http://www.knora.org/ontology/knora-base#standoffTagHasStart> ?standoffTagHasStart ;
        <http://www.knora.org/ontology/knora-base#standoffTagHasEnd> ?standoffTagHasEnd ;
        <http://www.knora.org/ontology/knora-base#standoffTagHasStartIndex> ?standoffTagHasStartIndex ;
        <http://www.knora.org/ontology/knora-base#standoffTagHasUUID> ?standoffTagHasUUID .
    }
    BIND(CONCAT(STR(?standoffTagHasStart), "|", STR(?standoffTagHasEnd), "|", STR(?standoffTagHasStartIndex), "|", ?standoffTagHasUUID) AS ?standoffNodeData)
  } UNION {
    VALUES ?resourceValueObject { <http://rdfh.ch/0801/beol/4auvM0v_TzKS4j2OVL8RiA/values/0W3GdDbtTgKfXR_jI62fHQ> <http://rdfh.ch/0801/beol/45jLFi29Rn-xl2wO8lCbkg/values/RM8btSjaQWepuwFnQ-s0Yw> <http://rdfh.ch/0801/beol/M7TWHhotSTiXGD9YSPZkNA/values/q8sJbBrnTSmHyswzK7vhLw> <http://rdfh.ch/0801/beol/-0tI3HXgSSOeDtkf-SA00w/values/GJQB9IHYTl2RhG4g4ru0YA> <http://rdfh.ch/0801/beol/LPclPw5MTm2jS_W7OE2ktw/values/gylS9v1hTfWH6_u6HfgU0Q> <http://rdfh.ch/0801/beol/Ou0NrqS8T-WY7lHTbkUKww/values/7iAa9ZXLTAqLCBp5yNKh3Q> <http://rdfh.ch/0801/beol/N80rnvmsTeuTF1QnLfdC2w/values/xbee-ZCRTE2XN6JURHcUqg> <http://rdfh.ch/0801/beol/6vuGLYDSQfqUBW8LOmMoaw/values/omx8e3xMSMKfUl3SAKJVow> <http://rdfh.ch/0801/beol/F6gjeBItSeS2bDGzmBe41w/values/lgKU4Hu9QyisG6TTwnr62g> <http://rdfh.ch/0801/beol/2PeTcAP2R0uyhxKMR163pA/values/NPWvNLAKTz-OOPbrXaH6Sg> <http://rdfh.ch/0801/beol/M8UMKPYOSqiLG4oqunxr-Q/values/mgC-KayaSYGHp6oYWqu2mQ> <http://rdfh.ch/0801/beol/QaDxYckoSY-Mu5tU8Gbz4w/values/QnP-RUaKTRWGzEayii8BfA> <http://rdfh.ch/0801/beol/5V0UQ8nyQquHXk5rdiWJ-Q/values/dBMQsL-IQNeIhuqke2Vw4Q> <http://rdfh.ch/0801/beol/0VE-1W0XRB-vzmZV2aemnA/values/SSMaDrOoRXSUaNMXQFEfZQ> <http://rdfh.ch/0801/beol/FiFZoNxOTvumQJSLonYp0A/values/q7NPps2UTIunu5j1WTLb7w> <http://rdfh.ch/0801/beol/CyrFhLONSVqyn0hMe1PyAA/values/zwa__RbFTUG9k3Pxg4R4gQ> <http://rdfh.ch/0801/beol/QewXbcXNTZOzPuKyEirsiw/values/aZWbXo1VSxOcByMm7bL79Q> <http://rdfh.ch/0801/beol/OIUG-7dnReCskHcKecGU6w/values/IINpkmISQr2ecIwRs9IRaA> <http://rdfh.ch/0801/beol/PJDKzIXlRomYrcsJ3bvDOg/values/fdblBD43SYelpS_Xr1mUxQ> <http://rdfh.ch/0801/beol/IBu4vDCGSTm4G3TNsOOtyA/values/epaW8kuXT4qsc4ZuFE9ZGw> <http://rdfh.ch/0801/beol/M-ZlPGzRTV2Qjye8ynw1QQ/values/E3yHIBWtQ8SGjfuqQx5HDw> <http://rdfh.ch/0801/beol/4Wfnmh7uQ522AH9OWEbygA/values/Sss4_gFqTvefnkmI027TnQ> <http://rdfh.ch/0801/beol/F_aBo9KCSJyRIGXXTLta5w/values/LsHcTGTVTIevf9K6NHOT5Q> <http://rdfh.ch/0801/beol/8b6a0T01T4iOODgOGDDPcQ/values/xYV-JzpdSU-JDNrdGXl_9w> <http://rdfh.ch/0801/beol/3tct2_3BRLmnSoA45OHBAA/values/VcA57tXxSbKn8muHHbyYCg> }
    GRAPH <http://www.ontotext.com/explicit> {
      ?resourceValueObject <http://www.knora.org/ontology/knora-base#valueHasStandoff> ?standoffNode .
    }
    GRAPH <http://www.ontotext.com/explicit> {
      ?standoffNode ?standoffProp ?standoffValue .
    }
      FILTER (?standoffProp != <http://www.knora.org/ontology/knora-base#standoffTagHasStart> && ?standoffProp != <http://www.knora.org/ontology/knora-base#standoffTagHasEnd> && ?standoffProp !=
        <http://www.knora.org/ontology/knora-base#standoffTagHasStartIndex> && ?standoffProp != <http://www.knora.org/ontology/knora-base#standoffTagHasUUID>)
    }
}
```

This returns 31,326 statements instead of 52,000. Every standoff tag is returned like this:

```turtle
<http://rdfh.ch/0801/beol/3tct2_3BRLmnSoA45OHBAA/values/VcA57tXxSbKn8muHHbyYCg/standoff/KQIy-qhnR52L0d1EsPwM1w>
  a <http://www.knora.org/ontology/standoff#StandoffItalicTag> ;
  <http://www.knora.org/ontology/knora-base#standoffNodeHasData> "2847|2851|78|f5972eac-69f3-4baf-a8f3-71eedc2387fb" ;
  <http://www.knora.org/ontology/knora-base#standoffTagHasStartParent> <http://rdfh.ch/0801/beol/3tct2_3BRLmnSoA45OHBAA/values/VcA57tXxSbKn8muHHbyYCg/standoff/MC9n8S2eS5qRTGjI4OaFhQ> .
```

It doesn't seem to be any slower; the GraphDB workbench still says it took 0.1 seconds.

**Comment by tobiasschweizer** [comment]
Cool! Just make sure the separator is a character that cannot possibly occur in the Iris :-)

**Comment by benjamingeer** [comment]
We can use the usual Unicode separator, this is just for illustration.

**Comment by tobiasschweizer** [comment]
Aha :-) Great!

**Comment by benjamingeer** [comment]
Actually not even necessary, because there are no IRIs in the concatenated string, just numbers and a UUID.

**Comment by benjamingeer** [comment]
We should also Base64-encode those UUIDs for storage in the triplestore, they'd be much shorter.

**Comment by benjamingeer** [comment]
Current plan: don't return any standoff markup, because whatever tag we return, its contents could be too long.

Ideally, if you used `knora-api:match`, we'd like to return some context (say, 100 characters) of the terms you searched for. The problem I see is, how do we know which text value contained the terms you searched for? What if your search matched a resource that has several different text values?

For example, suppose you do this search:

```sparql
CONSTRUCT {
  ?res knora-api:isMainResource true .
  ?res ?prop1 ?textVal1 .
  ?res ?prop2 ?textVal2 .
} WHERE {
  ?res a example:book .
  ?res ?prop1 ?textVal1 .
  ?res ?prop2 ?textVal2 .
  FILTER((?prop1 = example:title || ?prop1 = example:description) && (?prop2 = example:author || ?prop2 = example:publisher))
  FILTER knora-api:match(?textVal1, "foo")
  FILTER knora-api:match(?textVal2, "bar")
}
```

Now suppose there's a matching resource with these values:

* `example:title "bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar bar foo"`
* `example:publisher  "Foo foo foo publishers of foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo foo bar books"`

How would we know to return the last 100 characters of each of these values, instead of the first 100 characters? In other words, how would we know that `foo` matched in the title, and `bar` matched in the publisher? We would have to analyse the query, which would be difficult.

My suggestion: always return the first 100 characters of each text value in each matching resource, and don't worry about locating the words that the user was searching for.

**Comment by benjamingeer** [comment]
Actually, when we process the results of the main query, we do know which value IRIs correspond to which variables, right? So the logic could be like this:

1. We know that the user was looking for `foo` in `?textVal1`, because of `FILTER knora-api:match(?textVal1, "foo"`.
2. For each value corresponding to `?textVal1`, find the first `foo`, and trim the `valueHasString` to show that `foo` in context.

Right?

**Comment by tobiasschweizer** [comment]
Yes, but we only know that the text value matched. We do not know which part of it did.

Could this be helpful in some way: https://github.com/dhlab-basel/Knora/issues/630#issuecomment-353546410?

**Comment by benjamingeer** [comment]
> Yes, but we only know that the text value matched. We do not know which part of it did.

Couldn't we just find the search term in the text in Scala?

**Comment by benjamingeer** [comment]
Could we just allow a subset of Lucene Parser Syntax, which we can easily convert to regular expressions? Would it be enough just to support `*`?

**Comment by benjamingeer** [comment]
Depends on #926.

**Comment by benjamingeer** [comment]
GraphDB's Lucene connector can return snippets of matching text: http://graphdb.ontotext.com/documentation/free/lucene-graphdb-connector.html#snippet-extraction

**Comment by benjamingeer** [comment]
We need to make the prequery return not only the IRI of the matching text value, but also the IRIs of any matching standoff tags.

**Comment by tobiasschweizer** [comment]
I tried to configure a connector for my repo but I did not succeed. Then I just tried with the test data and the examples provided in the docs and it worked:

- create a new repo with standard settings
- upload sample data from http://graphdb.ontotext.com/documentation/standard/lucene-graphdb-connector.html#usage -> sample data (insert as RDF text snippet in web console)
- create connector using the SPARQL provided here http://graphdb.ontotext.com/documentation/standard/lucene-graphdb-connector.html#using-the-create-command
- do a query using the connector and field name http://graphdb.ontotext.com/documentation/standard/lucene-graphdb-connector.html#snippet-extraction

Result:
<img width="1375" alt="screen shot 2019-02-22 at 4 31 40 pm" src="https://user-images.githubusercontent.com/6000023/53252519-66088a80-36bf-11e9-9177-3b9318f03f34.png">

So it is possible to get the Iri of the entity that matched **and** the matching text plus some context. By default, `<em>` tags are inserted to mark the match, but this is configurable.

I wonder if there is more information that could be returned like the start and end position of the match. With that information, the relevant standoff could be queried.

**Comment by benjamingeer** [comment]
> With that information, the relevant standoff could be queried.

Maybe you don't actually need all of the relevant standoff tags. Would it be good enough just to have the maths tags?

**Comment by benjamingeer** [comment]
Also, it's great to see that you can get **something**. :)

**Comment by tobiasschweizer** [comment]
> Would it be good enough just to have the maths tags?

I think that would solve our problem of rendering the LaTeX properly when doing a full text search. MathJax expects LaTeX to appear within delimiters and those are created from standoff math tags (there are not in the source, list nodes are an exception).

> Also, it's great to see that you can get **something**. :)

Yes, I also think this is great! :-)

Maybe I just ask Ontotext if there is more information that can be accessed.

**Comment by tobiasschweizer** [comment]
We should add a boolean parameter, e.g., `returnCompleteTextValues`, to `GravsearchRequestV2` so Gravsearch queries that are used internally (like the TEI export) could return complete standoff information. For Gravsearch queries submitted by the client not all standoff information would be returned.

**Comment by benjamingeer** [comment]
The same problem can occur when loading just one text value: #1293

**Comment by benjamingeer** [comment]
I have a crazy idea for fixing this, which doesn't require any special support from the triplestore, but would require us to return standoff as standoff in the API instead of as XML.

We could **never return standoff with the text value**. Instead, we would just return the plain text.

The standoff tags would have to be requested separately, and would always be paged, e.g. 25 tags per page. This would be easy because they're ordered by an integer index.

We could also return the "context" of tags that surround a particular point in the text.

**Comment by tobiasschweizer** [comment]
> I have a crazy idea for fixing this, which doesn't require any special support from the triplestore, but would require us to return standoff as standoff in the API instead of as XML.

I do not think that your idea is crazy. In fact, it is quite consistent with our LOD approach. We need a API format for standoff tags that is XML free. However, there should still be a possibility to query a text value as XML or HTML since not all clients can cope with complex LOD data.

I would like to try converting standoff tags into XML or HTML on the client side, e.g., in an Angular directive.

**Comment by benjamingeer** [comment]
> I do not think that your idea is crazy.

Maybe you're as crazy as I am. :)

> However, there should still be a possibility to query a text value as XML or HTML since not all clients can cope with complex LOD data.

The problem is how to support large documents in a way that's consistent with the API v2 design goal of keeping responses small and fast.

The goal of #1259 is to generate client libraries for different programming languages, so clients don't have to deal with the details of Knora API responses. Maybe these libraries could handle something like this as well, so clients wouldn't even need to know about it. If the client needs XML, the client library could handle getting the standoff (one page at a time) and then converting it to XML.

> I would like to try converting standoff tags into XML or HTML on the client side, e.g., in an Angular directive.

That would be great.

**Comment by tobiasschweizer** [comment]
I think we should give it a try quite soon. Also we should try to return the context in full text search results. Any news from Ontotext about the snippet extraction?

**Comment by benjamingeer** [comment]
We could also provide an "export to XML" feature that would work on the server, like the planned "export to CSV" (#919), and would handle large files.

> Any news from Ontotext about the snippet extraction?

Nope.

**Comment by gfoo** [comment]
don't really understand your discussion, but note that it is really useful to have the xml content of a RichText property in the resource response :-) Not sure to deal with an another endpoint will help.

**Comment by benjamingeer** [comment]
> don't really understand your discussion, but note that it is really useful to have the xml content of a RichText property in the resource response :-) Not sure to deal with an another endpoint will help.

The problem is that to convert the standoff to XML, we need **all** the standoff for the text value. This can be a huge amount of data, as you saw in #1293. We are trying to avoid getting huge amounts of data from the triplestore per request, to make requests faster and to reduce the load on the triplestore. Getting the standoff separately, in small pages of data, would help because it would make the requests smaller and faster.

There will still be a way to convert the markup to XML, but this will be done on the client instead of on the server.

**Comment by gfoo** [comment]
ok, I will prepare myself (psychologically) to change my mind ;-)

**Comment by tobiasschweizer** [comment]
@benjamingeer I think there should be several ways to get a text with markup from Knora, and one of them should definitely serve the markup in an XML-free standoff format. 

As for PR #1307: is your intention to always return standoff separately from texts or just offer a way to do so in the API? 

In order to solve the problem with the triplestore, couldn't Knora query for the standoff in several iterations? Then, the markup could still be returned in XML or HTML.

**Comment by benjamingeer** [comment]
> As for PR #1307: is your intention to always return standoff separately from texts or just offer a way to do so in the API? 

For now, offer a way to do so in the API, with XML remaining as the default. Later, when we have good client-side support for editing and displaying standoff and converting it to XML, I think the current server-side XML generation could be deprecated and eventually removed.

> In order to solve the problem with the triplestore, couldn't Knora query for the standoff in several iterations? Then, the markup could still be returned in XML or HTML.

Querying in several (perhaps many) iterations, then allowing the user to download the resulting (potentially huge) XML document, is what I was talking about in my [comment above](https://github.com/dhlab-basel/Knora/issues/913#issuecomment-482079199): an “export to XML” feature. The user would request the export and get a confirmation response immediately. They would later get a notification when the export was ready for download. This would require the batch-processing functionality planned in #1210.

**Comment by tobiasschweizer** [comment]
@benjamingeer I hope I will back to work on Wednesday, I am very interested in contributing to this PR. I think offering a way to get standoff through the API makes Knora truly LOD.

**Comment by benjamingeer** [comment]
> In order to solve the problem with the triplestore, couldn't Knora query for the standoff in several iterations? Then, the markup could still be returned in XML or HTML.

I have an idea about how to do this, but I'll try it in a separate PR.

**Comment by benjamingeer** [comment]
> > In order to solve the problem with the triplestore, couldn't Knora query for the standoff in several iterations? Then, the markup could still be returned in XML or HTML.
>
> I have an idea about how to do this, but I'll try it in a separate PR.

Now it looks like I'll have to do it sooner rather than later.

**Comment by tobiasschweizer** [comment]
> Now it looks like I'll have to do it sooner rather than later.

maybe this is the appropriate occasion for the rocket emoticon
